### ISSUE TITLE
[New Concept Docs]: Functional Tools in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3107]
This issue describes how to implement the `functional tools` in Python  **concept docs**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3108)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach an understanding/use of core `functional tools` (e.g,  **`map()`**, **`filter()`**, and **`functools.reduce()`**  in Python. 

<br>

## 💡   Learning objectives

- Understand/use the `built-in map()` function.
  - constructing a `lambda`  or callback function used as the argument.
  - comparisons to `set`, `dict` or `list` comprehensions
  - when and when not to use - considering performance & readability
- Understand/use the `built-in filter()` function.
  - constructing a `lambda`  or callback function used as the argument.
  - comparisons to `set`, `dict` or `list` comprehensions
  - when and when not to use - considering performance & readability
- Understand/use the `functools.reduce()` function
  - constructing a `lambda`  or callback function used as the argument.
  - using `initializer` as a "first call" or to guard against TypeErrors when the passed `iterable` is empty.
  - comparisons to `sum()`, `min()`, `max()`, `any()`, `all()`, `math.prod()` and to `intertools.accumulate()`
  - when and when not to use - considering performance & readability

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to open</em>)</summary>
<br>

- `comprehensions`
- `comprehensions` in `lambdas`
- `map()`, `filter()` or `functools.reduce()` within a `comprehension`
- `functools` beyond `functools.reduce()`(_this will get its own exercise_)
- `generators`
-  `map()`, `filter()` or `functools.reduce()` within a `generator expression`
- using an `assignment expression` or "walrus" operator (`:=`) in a `lambda` (_walrus in general is OK if it is explained, just not inside a lambda_)

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts & Related Concepts Covered</b> (<em>click to open</em>)</summary>
<br>

- functional tools in python
- `built-ins`
  - `any()` & `all()`
  - `sum()`
  - `min()` & `max()`
  - `map()`
  - `filter()`
- `functools.reduce()`
- `itertools.accumulate()`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs  (<em>click to open</em>)</summary>
<br>

- `basics`
- `bools`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `iteration`
-  `lambdas`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>  

## 📚  Resources to refer to

<details>
<summary><b>Resources</b>  (<em>click to open</em>)</summary>
<br>

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs Tutorial: Lambda Expressions](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [built-ins: Python Docs](https://docs.python.org/3/library/functions.html)
  - [`map()`](https://docs.python.org/3/library/functions.html#map) & [`filter()`](https://docs.python.org/3/library/functions.html#filter)
  - [`any()`](https://docs.python.org/3/library/functions.html#any) & [`all()`](https://docs.python.org/3/library/functions.html#all)
  - [`min()`](https://docs.python.org/3/library/functions.html#min) & [`max()`](https://docs.python.org/3/library/functions.html#max)
  - [`sum()`](https://docs.python.org/3/library/functions.html#sum) 
- [`itertools.accumulate()`](https://docs.python.org/3/library/itertools.html#itertools.accumulate)
- [`functools.reduce()`](https://docs.python.org/3/library/functools.html#functools.reduce)
- [Real Python: Functional Programming in Python: When and How to Use it](https://realpython.com/python-functional-programming/)
- [Real Python : Python's map(): Processing Iterables Without a Loop](https://realpython.com/python-map-function/)
- [Real Python: Python's reduce(): From Functional to Pythonic Style](https://realpython.com/python-map-function/)

</details>
<details>
<summary><b>Additional Articles</b>  (<em>click to open</em>)</summary>
<br>

- [Learn by Example:  Python Lambda Function](https://www.learnbyexample.org/python-lambda-function/)
- [Python Docs:  Functional Programming HOWTO](https://docs.python.org/3/howto/functional.html)

</details>

<br>

##  📁  Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b> (<em>click to collapse</em>)</summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
